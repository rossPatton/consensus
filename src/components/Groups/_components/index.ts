import loadable from '@loadable/component';

export const Group = loadable(() =>
  import(/* webpackChunkName: "Group Sub Component" */'./Group'),
);
