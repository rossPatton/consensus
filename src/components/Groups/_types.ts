export type tProps = {
  asList?: boolean,
  count?: number,
  isEditable?: boolean,
  hoverIndex?: number,
  groups: ts.group[],
  roles: ts.roleMap[],
  showCategory?: boolean,
  showLocation?: boolean,
  showType?: boolean,
  showPending?: boolean,
};

export type tComponentProps = tProps & {
  leaveOrg: (ev: React.MouseEvent<HTMLButtonElement>, groupId: number) => void,
  pendingOrgs: ts.group[],
  setHover: (hoverIndex: number | null, groupType?: ts.role) => void,
};

export type tContainerProps = tProps & {
  deleteOrgByUserIdDispatch: (query: ts.deleteUserByGroupIdQuery) => ts.thunkPayload<ts.group>,
};

export type tState = {
  groupType: ts.role,
  hoverIndex: number | null,
};
