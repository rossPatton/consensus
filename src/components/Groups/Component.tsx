import cx from 'classnames';
import _ from 'lodash';
import React, {memo} from 'react';

import {Group} from './_components';
import {tComponentProps} from './_types';

export const GroupsComponent = memo((props: tComponentProps) => (
  <ul className={cx({'flex flex-wrap': !props.asList})}>
    {props.groups.length > 0
      && props.pendingOrgs.length > 0
      && (
        <li className="font-bold mb-2 pb-2 border-b-1">
          Current group memberships
        </li>
      )}
    {props.groups.map((group, i) => (
      <Group
        {...props}
        key={group.handle}
        group={group}
        index={i}
      />
    ))}
    {props.pendingOrgs.length > 0 && (
      <>
        <li className="font-bold mb-2 pb-2 border-b-1">
          Pending group memberships
        </li>
        {props.pendingOrgs.map((pendingOrg, i) => (
          <Group
            {...props}
            index={i}
            key={pendingOrg.id}
            group={pendingOrg}
          />
        ))}
      </>
    )}
  </ul>
));
