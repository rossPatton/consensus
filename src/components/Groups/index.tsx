import _ from 'lodash';
import React, {PureComponent} from 'react';
import {connect} from 'react-redux';

import {Paginate} from '~app/containers';
import {deleteOrgByUserId} from '~app/redux';

import {tContainerProps, tState} from './_types';
import {GroupsComponent} from './Component';

class GroupsContainer extends PureComponent<tContainerProps, tState> {
  state = {
    hoverIndex: null as number | null,
    // we sometimes split the groups into 2, by membership vs pending
    // short of splitting this entire thing into 2 nearly identical components
    // (which might be the best approach honestly)
    // i'm using this to only render the hover state if the type matches
    groupType: 'member' as ts.role,
  };

  // ie, return an array of only groups that pending
  filterNonPending = () => {
    return this.props.groups.filter(group => {
      const roleMap = _.find(this.props.roles, r => r.groupId === group.id) || {};
      const {role} = roleMap as ts.roleMap;
      return role === 'pending';
    });
  }

  // only render groups that the user is a member of
  filterPending = () => {
    return this.props.groups.filter(group => {
      const roleMap = _.find(this.props.roles, r => r.groupId === group.id) || {};
      const {role} = roleMap as ts.roleMap;
      return role !== 'pending';
    });
  }

  leaveOrg = (ev: React.MouseEvent<HTMLButtonElement>, groupId: number) => {
    ev.preventDefault();
    if (groupId) {
      this.props.deleteOrgByUserIdDispatch({groupId});
    }
  }

  setHover = (hoverIndex: number | null = null, groupType: ts.role) =>
    this.setState({
      groupType,
      hoverIndex,
    });

  render() {
    const {
      asList,
      count,
      isEditable,
      groups: originalGroups,
      roles,
      showCategory,
      showLocation,
      showPending = false,
      showType,
    } = this.props;

    if (!originalGroups
      || (originalGroups instanceof Array && originalGroups.length === 0)) {
      return null;
    }

    // if we're rendering groups based on relation to the user, then we want to
    // be able to potentially split by user role, pending or not pending
    let pendingOrgs = [] as ts.group[];
    if (showPending) {
      pendingOrgs = this.filterNonPending();
    }

    return (
      <Paginate
        count={count}
        items={this.filterPending()}
        render={(groupsToRender: ts.group[]) => (
          <GroupsComponent
            {...this.state}
            asList={asList}
            isEditable={isEditable}
            leaveOrg={this.leaveOrg}
            groups={groupsToRender}
            pendingOrgs={pendingOrgs}
            roles={roles}
            setHover={this.setHover}
            showCategory={showCategory}
            showLocation={showLocation}
            showType={showType}
          />
        )}
      />
    );
  }
}

const mapStateToProps = (store: {roles: ts.thunk<ts.roleMap[]>}) => ({
  roles: store.roles.data,
});

const mapDispatchToProps = (dispatch: Function) => ({
  deleteOrgByUserIdDispatch: (query: ts.deleteUserByGroupIdQuery) =>
    dispatch(deleteOrgByUserId(query)),
});

const Group = connect(
  mapStateToProps,
  mapDispatchToProps,
)(GroupsContainer);

export default Group;
