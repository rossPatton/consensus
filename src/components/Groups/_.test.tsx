import React from 'react';
import { Provider } from 'react-redux';
import render from 'react-test-renderer';
import configureStore from 'redux-mock-store';

const mockStore = configureStore();
import Groups from '.';

describe('Groups component', () => {
  it('renders without crashing', () => {
    const store = mockStore({
      roles: [{groupId: 100, role: 'facilitator'}],
    });

    render.create((
      <Provider store={store}>
        <Groups groups={[]} />
      </Provider>
    ));
  });
});
