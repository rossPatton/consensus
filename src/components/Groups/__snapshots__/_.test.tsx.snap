// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Groups group component onClick 1`] = `null`;

exports[`components/Groups renders container 1`] = `null`;

exports[`components/Groups renders container with no groups 1`] = `null`;

exports[`components/Groups renders editable group component with location 1`] = `null`;

exports[`components/Groups renders editable pending group component as list 1`] = `null`;

exports[`components/Groups renders groups component 1`] = `
<ul
  className="flex flex-wrap"
>
  <li
    className="font-semibold mb-2"
  >
    Current group memberships
  </li>
  <li
    className="font-semibold mb-2"
  >
    Pending group memberships
  </li>
</ul>
`;

exports[`components/Groups renders individual group component with type 1`] = `null`;

exports[`components/Groups renders individual pending group component with category 1`] = `null`;
